src/Data/JSON/Directory.hs

```diff
@@ -28,32 +34,58 @@ data ModifiedWhileReading = ModifiedWhileReading FilePath
 
 instance Exception ModifiedWhileReading
 
+data NoRuleFor = NoRuleFor FilePath
+    deriving Show
+
+instance Exception NoRuleFor
+
+data Rule = Rule
+    { predicate :: FilePath -> Bool
+    , jsonKey   :: FilePath -> Text
+    , parser    :: FilePath -> IO (IResult Value)
+    }
+
+jsonRule :: Rule
+jsonRule = Rule
+    { predicate = isSuffixOf ".json"
+    , jsonKey   = Text.pack . takeBaseName
+    , parser    = idecodeFileStrict
+    }
+
+textRule :: Rule
+textRule = Rule
+    { predicate = const True
+    , jsonKey   = Text.pack . takeFileName
+    , parser    = fmap (ISuccess . String) . Text.readFile
+    }
+
+defaultRules :: [Rule]
+defaultRules = [jsonRule, textRule]
+
 data EntryType
     = Directory
-    | JSON
-    | TextFile
+    | File (FilePath -> IO (IResult Value))
```

Comment:
This looks like the most flexible API to me. Since we need some result type, I don't see any problem with reusing this one.

The docs might just need a note that when returing `IError` the JSONPath can be used to to provide context for the error relative to that file, since we use `addContext` to automatically prefix the context for that file.